client/state/data-layer/http-data.ts

```diff
@@ -0,0 +1,345 @@
+
+type DataId = string;
+
+export const httpData = new Map< DataId, Resource >();
+export const listeners = new Set();
+
+export const empty: Readonly< Resource > = Object.freeze( {
+	state: DataState.Uninitialized,
+	data: undefined,
+	error: undefined,
+	lastUpdated: -Infinity,
+	pendingSince: undefined,
+} );
+
+export const getHttpData = ( id: DataId ) => httpData.get( id ) || empty;
+
+export const subscribe = ( f: Function ): ( () => void ) => {
+	listeners.add( f );
+
+	return () => void listeners.delete( f );
+};
+
+export const updateData = ( id: DataId, state: DataState, data: any ) => {
+	const lastUpdated: TimestampMS = Date.now();
+	const item = httpData.get( id ) || empty;
+
+	// We could have left out the keys for
+	// the previous properties if they didn't
+	// exist but I wanted to make sure we can
+	// get our hidden classes to optimize here.
+	switch ( state ) {
+		case DataState.Failure:
+			return httpData.set( id, {
+				state,
+				data: item.data,
+				error: data,
+				lastUpdated: item.lastUpdated,
+				pendingSince: undefined,
+			} );
+
+		case DataState.Pending:
+			return httpData.set( id, {
+				state,
+				data: item.data,
+				error: undefined,
+				lastUpdated: item.lastUpdated,
+				pendingSince: lastUpdated,
+			} );
+
+		case DataState.Success:
+			return httpData.set( id, {
+				state,
+				data,
+				error: undefined,
+				lastUpdated,
+				pendingSince: undefined,
+			} );
+	}
+};
+
+export const update = ( id: DataId, state: DataState, data?: any ) => {
+	const updated = updateData( id, state, data );
+
+	listeners.forEach( f => f() );
+
+	return updated;
+};
+
+const fetch = ( action: HttpDataAction ) => {
+	update( action.id, DataState.Pending );
+
+	return [
+		{
+			...action.fetch,
+			onSuccess: action,
+			onFailure: action,
+			onProgress: action,
+		},
+		{ type: HTTP_DATA_TICK },
+	];
+};
+
+const onError = ( action: HttpDataAction, error: any ) => {
+	update( action.id, DataState.Failure, error );
+
+	return { type: HTTP_DATA_TICK };
+};
+
+type SuccessfulParse = [undefined, ReturnType< ResponseParser >];
+type FailedParse = [Error, undefined];
+type ParseResult = SuccessfulParse | FailedParse;
+
+/**
+ * Transforms API response data into storable data
+ * Returns pairs of data ids and data plus an error indicator
+ *
+ * [ error?, [ [ id, data ], [ id, data ], … ] ]
+ *
+ * @example:
+ *   --input--
+ *   { data: { sites: {
+ *     14: { is_active: true, name: 'foo' },
+ *     19: { is_active: false, name: 'bar' }
+ *   } } }
+ *
+ *   --output--
+ *   [ [ 'site-names-14', 'foo' ] ]
+ *
+ * @param data - input data from API response
+ * @param fromApi - transforms API response data
+ * @return output data to store
+ */
+const parseResponse = ( data: any, fromApi: ResponseParser ): ParseResult => {
+	try {
+		return [ undefined, fromApi( data ) ];
+	} catch ( error ) {
+		return [ error, undefined ];
+	}
+};
+
+const onSuccess = ( action: HttpDataAction, apiData: any ) => {
```

Comment:
during our conversation about this piece we were looking for the type  `unknown` without realizing it.
with `unknown` we cannot make any use of the data because it strips it down to a true black block, while `any` allow _any_ use of the data

```ts
function test( open: any, closed: unknown ) {
	open.run(); // will not cause an error because `any` satisfies `{ run: Function }`
	closed.run(); // Type Error because object is of type `unknown`
}
```

there are a few places in here we should replace `any` with `unknown`

Reply:
I played with this a bit. A few notes I found helpful:

```ts
// Happy with whatever, it's any 🤷
const f = (x: any) => {
  x.boom = 1; // ✅
  numId( x ); // ✅
  anyId( x ); // ✅
}

// `unknown` is limiting,
// but we can use it as `any`
const g = (x: unknown) => {
  x.boom = 1; // ⛔️
  numId( x ); // ⛔️
  anyId( x ); // ✅
}

const numId = (x: number): number => x
const anyId = (x: any): any => x
```

I was thrown because I thought `unknown` would be a sort of dead-end type, but it does satisfy `any` which initially surprised me. I'll review how that fits into this PR, but y'all should feel free to make those changes as well.

Reply:
> it does satisfy `any` which initially surprised me

well remember that `any` is fulfilled by _any_ data, so `unknown` is included. this is a big reason to avoid `any` everywhere except where we have no choice